compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -682,23 +682,27 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             };
 
             let encl_item_id = self.tcx.hir().get_parent_item(expr.hir_id);
-            let encl_item = self.tcx.hir().expect_item(encl_item_id);
 
-            if let hir::ItemKind::Fn(..) = encl_item.kind {
-                // We are inside a function body, so reporting "return statement
-                // outside of function body" needs an explanation.
+            // Somewhat confusingly, get_parent_item() does not necessarily return an
+            // item -- it can also return a Foreign-/Impl-/TraitItem or a Crate (see
+            // issue #86721). If it does, we still report the same error.
+            if let Some(hir::Node::Item(encl_item)) = self.tcx.hir().find(encl_item_id) {
```

Comment:
IIUC this doesn't cover associated/trait functions, does it?

So here's another idea I have: instead of matching on all kinds of body owner, can we check that the parent is a body owner (`tcx.hir().maybe_body_owned_by(encl_item_id).is_some()`) and if it is, get its span with `tcx.hir().span(encl_item_id)`?

Reply:
Thanks for the suggestion, but I don't think that this is what we want here. The `if` block is responsible for adding the following explanatory note to the error message:
```
error[E0572]: return statement outside of function body
 --> test.rs:2:10
  |
1 | / fn foo() {
2 | |     [(); return];
  | |          ^^^^^^ the return is part of this body...
3 | | }
  | |_- ...not the enclosing function body
```
Without it, or if the enclosing item is not a function, the same error is emitted, just without the extra note:
```
error[E0572]: return statement outside of function body
 --> test.rs:1:25
  |
1 | const N: usize = { [(); return]; };
  |                         ^^^^^^
```
The intention is that the note should explain why the `return` is outside of a function body, if it _appears_ to be inside one. But it only makes sense to issue such a note if the enclosing item actually _is_ a function.

You are right in that the check should also cover functions in trait and impl items, which I have fixed now.

Reply:
Okay, sounds good.